File: IPaURLResourceUI/IPaURLResourceUI/IPaServerAPI.m
//
//  IPaServerAPI.m
//  IPaURLResourceUI
//
//  Created by IPa Chen on 2015/2/14.
//  Copyright (c) 2015年 IPaPa. All rights reserved.
//

#import "IPaServerAPI.h"

@implementation IPaServerAPI
{
    NSURLSessionDataTask *currentAPITask;
}
- (instancetype) init
{
    return nil;
}
- (instancetype)initWithResourceUI:(IPaURLResourceUI*)resourceUI
{
    self = [super init];
    self.resourceUI = resourceUI;
    return self;
}
- (void) api:(NSString *)api Method:(NSString *)method Param:(NSDictionary *)aParam success:(IPaURLResourceUISuccessHandler)success
     failure:(void (^)(NSError *error))failure {
    if (currentAPITask != nil && currentAPITask.state == NSURLSessionTaskStateRunning) {
        [currentAPITask cancel];
    }
    method = [method uppercaseString];
    if ([method isEqualToString:@"GET"]) {
        currentAPITask = [self.resourceUI apiGet:api param:aParam onComplete:success failure:failure];
    }
    else if ([method isEqualToString:@"POST"]) {
        currentAPITask = [self.resourceUI apiPost:api param:aParam onComplete:success failure:failure];
    }
    
}
-(void) api:(NSString*)api putJSON:(id)jsonObject success:(IPaURLResourceUISuccessHandler)success
    failure:(void (^)(NSError *error))failure {
    NSError *jsonError;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:jsonObject options:0 error:&jsonError];
    if (jsonError != nil) {
        if (failure) {
            failure(jsonError);
        }
        return;
    }
    
    
    if (currentAPITask != nil && currentAPITask.state == NSURLSessionTaskStateRunning) {
        [currentAPITask cancel];
    }
    
    currentAPITask = [self.resourceUI apiPut:api contentType:@"application/json" postData:jsonData onComplete:success failure:failure];
}

-(void) api:(NSString*)api postJSON:(id)jsonObject success:(IPaURLResourceUISuccessHandler)success
    failure:(void (^)(NSError *error))failure {
    NSError *jsonError;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:jsonObject options:0 error:&jsonError];
    if (jsonError != nil) {
        if (failure) {
            failure(jsonError);
        }
        return;
    }
    
    
    if (currentAPITask != nil && currentAPITask.state == NSURLSessionTaskStateRunning) {
        [currentAPITask cancel];
    }
    
    currentAPITask = [self.resourceUI apiPost:api contentType:@"application/json" postData:jsonData onComplete:success failure:failure];
}
@end

File: IPaURLResourceUI/IPaURLResourceUI/IPaServerAPI.swift
//
//  IPaServerAPI.swift
//  IPaURLResourceUI
//
//  Created by IPa Chen on 2015/6/14.
//  Copyright (c) 2015年 IPaPa. All rights reserved.
//

import Foundation

class IPaServerAPI :NSObject {
    let resourceUI:IPaURLResourceUI
    var currentAPITask:URLSessionDataTask?
    init(resourceUI:IPaURLResourceUI) {
        self.resourceUI = resourceUI
    }
    func apiPerform(_ api:String,method:String,param:[String:AnyObject]?,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) {
        
        if let task = currentAPITask {
            if task.state == .running {
                task.cancel()
            }
        }

        let aMethod = method.uppercased()
        if aMethod == "GET" {
            currentAPITask = resourceUI.apiGet(api, param:param, complete:{
                response,responseObject in
                self.currentAPITask = nil
                complete(response,responseObject)
                },failure:{
                    error in
                self.currentAPITask = nil
                    failure(error)
            })
        }
        else if aMethod == "POST" {
            currentAPITask = resourceUI.apiPost(api, param:param,  complete:{
                response,responseObject in
                self.currentAPITask = nil
                complete(response,responseObject)
                },failure:{
                    error in
                    self.currentAPITask = nil
                    failure(error)
            })
        }

    }
    func apiPut(_ api:String, json:AnyObject,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) {
        var jsonError:NSError?
        var jsonData:Data?
        do {
            jsonData = try JSONSerialization.data(withJSONObject: json, options: JSONSerialization.WritingOptions())
        } catch let error as NSError {
            jsonError = error
            jsonData = nil
        }
        
        if let error = jsonError {
            
            failure(error)
            
            return
        }
        if let task = currentAPITask {
            if task.state == .running {
                task.cancel()
            }
        }
        
        currentAPITask = resourceUI.apiPut(api, contentType: "application/json", postData: jsonData!,  complete:{
            response,responseObject in
            self.currentAPITask = nil
            complete(response,responseObject)
            },failure:{
                error in
                self.currentAPITask = nil
                failure(error)
        })


    }
    func apiPost(_ api:String,json:AnyObject,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) {
        var jsonError:NSError?
        var jsonData:Data?
        do {
            jsonData = try JSONSerialization.data(withJSONObject: json, options: JSONSerialization.WritingOptions())
        } catch let error as NSError {
            jsonError = error
            jsonData = nil
        }
        
        if let error = jsonError {
            
            failure(error)
            
            return
        }
        if let task = currentAPITask {
            if task.state == .running {
                task.cancel()
            }
        }
        
        currentAPITask = resourceUI.apiPost(api, contentType: "application/json", postData: jsonData!,  complete:{
            response,responseObject in
            self.currentAPITask = nil
            complete(response,responseObject)
            },failure:{
                error in
                self.currentAPITask = nil
                failure(error)
        })
    }
}
